libutils/sequence.c

```diff
@@ -449,3 +450,87 @@ int SeqStringLength(Seq *seq)
 
     return total_length;
 }
+
+// TODO: These static helper functions could be (re)moved
+static bool HasNulByte(const char *str, size_t n)
+{
+    for (int i = 0; i<n; ++i)
+    {
+        if (str[i] == '\0')
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+static long GetLengthPrefix(const char *data)
+{
+    if (HasNulByte(data, 10))
+    {
+        return -1;
+    }
+    if (!isdigit(data[0]))
+    {
+        return -1;
+    }
+    size_t length;
+    int ret = sscanf(data, "%10zd", &length);
+    if (ret != 1)
+    {
+        // incorrect number of items scanned
+        return -1;
+    }
+    return (long)length;
+}
+
+char *SeqStringSerialize(Seq *seq)
+{
+    assert(seq != NULL);
+    size_t length = SeqLength(seq);
+    Writer *w = StringWriter();
+    for (int i = 0; i < length; ++i)
+    {
+        const char *s = SeqAt(seq, i);
+        WriterWriteF(w, "%-9ld %s\n", strlen(s), s);
+    }
+    return StringWriterClose(w);
+}
+
+Seq *SeqStringDeserialize(const char *const serialized)
+{
+    const int PREFIX_LENGTH = 10;
+    Seq *seq = SeqNew(512, free);
+    const char *src = serialized;
+    while (src[0] != '\0')
+    {
+        // Read length prefix first
+        long length = GetLengthPrefix(src);
+
+        // Do validation here, this could be optimized by
+        // eliminating the extra pass (HasNulByte())
+        if (length < 0
+            || HasNulByte(src + PREFIX_LENGTH, length + 1)
+            || src[PREFIX_LENGTH - 1] != ' '
+            || src[PREFIX_LENGTH + length] != '\n')
+        {
+            SeqDestroy(seq);
+            return NULL;
+        }
+
+        // Allocate the buffer which will be inserted into Sequence
+        char *new_str = xmalloc(length+1); // +1 for NUL termination
+
+        // Advance the src pointer
+        src += PREFIX_LENGTH;
+
+        // Copy, NUL-terminate and add to Sequence
+        memcpy(new_str, src, length);
+        new_str[length] = '\0';
```

Comment:
I think you could use ``strndup()`` instead of the ``xmalloc(); memcpy(); new_str[length] = '\0'`` here.